### PR TITLE
PolyfaceQuery.sweepLineStringToFacets: enhancement to (now-deprecated) PolyfaceQuery.sweepLinestringToFacetsXYReturnLines

### PR DESCRIPTION
**Pull request by EarlinLutz**
New method PolyfaceQuery.sweepLineStringToFacets is an enhancement to behavior of PolyfaceQuery.sweepLinestringToFacetsXYReturnChain and PolyfaceQuery.sweepLinestringToFacetsXYReturnChains.  It provides
  * options to select among outputs on forward, side, and rear-facing facets.
  * an angle tolerance for when facets are considered "side"
  * options to select between segment and chain output.
  * Optional input of sweep direction (other than Z direction draping)
 * Two prior m
* https://github.com/iTwin/itwinjs-core/issues/5507 observes unexpected linework returned from PolyfaceQuery.sweepLinestringToFacetsXYReturnChain.
* for the test case given, the output described as zingers is correct.
  * these lines are on "almost vertical" facets which the caller was unaware of or thought were true vertical.
  * the output is correct
  * a second mesh that has an almost-identical facet which truly is vertical.  No output is generated on that facet.  The code is behaving "properly" at that point -- "Works as Designed" but there should be more consistent treatment so callers can reliably receive outputs for front, side and rear facing facets.

The specific changes introduced are:
 * New method PolyfaceQuery.sweepLineStringToFacets which implements all the options
 * New class SweepLineStringToFacetsOptions which conveys input options described above.
 * Angle methods
   * static Angle.createSmallAngle () -- return an angle object with the system-wide small angle.
   * Angle.isMagnitudeLessThanOrEqual (other) -- angle comparison method.
  * File polyface\multiclip\SweepLineStringToFacetContext.ts is an internal context to support drape methods.
     * Existing context SweepLineStringToFacetContext will someday be deprecated.  It is still used by various other drape methods (e.g. asynch methods that still use it and will continue to exhibit the behavior in 5507.)
     * New context (in same file) ClipSweptLineStringContext implements the new feature "sweep in any direction".
       * per-face classification (forward, side, rear) is implemented in the calling code in PolyfaceQuery.
 

An implementation detail:
* the deprecated methods (and their support context) used this logic for a facet and a segment to drape
  * construct a (skewed) coordinate frame with its UV plane in the polygon plane and W vector in the sweep direction (fixed at global Z.
  * transform both the polygon and the segment-to-drape into that frame
  * do polygon clip in that UV plane.
  * This logic fails when the sweep direction is in the plane of the polygon.
  * Correcting this failure is a main goal of this PR
* The new method uses this logic which does forward, side, and rear polygon clips cleanly
  * create a plane whose normal is perpendicular to (a) the drape segment and (b) the sweep direction
  * intersect the polygon with this plane.

Hence there is a reverse of roles:
  * previously had a "line polygon" clip within a plane
  * now have a "polygon-crosses-plane" clip.

### CODE REVIEW

File: docs/changehistory/NextVersion.md [hunk 1]
```diff
@@ -69,3 +71,39 @@ Here are those result sections assembled into `RuledSweep` solids and then facet
 ### Swept surface constructions
 
 The constructors for swept surfaces ([LinearSweep]($core-geometry), [RotationalSweep]($core-geometry), [RuledSweep]($core-geometry)) now allow [CurvePrimitive]($core-geometry) input. Internally, the curve is promoted to a `CurveChain` with one member.
+
+### Draping (sweeping) a linestring onto a mesh
```
Comment (pmconne) [inline docs/changehistory/NextVersion.md:75]:
```suggestion
### Sweeping a linestring to facets
```

File: docs/changehistory/NextVersion.md [hunk 2]
```diff
@@ -13,9 +13,11 @@ Table of contents:
   - [Renderer, editor and category on calculated properties](#renderer-editor-and-category-on-calculated-properties)
 - [Electron 25 support](#electron-25-support)
 - [Geometry](#geometry)
-  - [Sweeping a section to a sequence of planes](#sweeping-a-section-to-a-sequence-of-planes)
+  - [Sweeping a section to a sequence of planes]
+  - [Sweeping a linestring to facets]
```
Comment (pmconne) [inline docs/changehistory/NextVersion.md:17]:
```suggestion
  - [Sweeping a section to a sequence of planes](#sweeping-a-section-to-a-sequence-of-planes)
  - [Sweeping a linestring to facets](#sweeping-a-linestring-to-facets)
```

Comment (pmconne) [reply docs/changehistory/NextVersion.md:17]: These are supposed to be links. The first one was a link prior to your edit.

**Review by pmconne — CHANGES_REQUESTED** [review]

File: docs/changehistory/NextVersion.md (hunk 2 above)
Comment (pmconne) [reply docs/changehistory/NextVersion.md:17]: Unless you object to proposed changes, you can just use the "Commit suggestion" buttons to apply them.

File: core/geometry/src/geometry3d/Angle.ts [hunk 3]
```diff
@@ -61,6 +61,12 @@ export class Angle implements BeJSONFunctions {
   public static createRadians(radians: number): Angle {
     return new Angle(radians);
   }
+  /**
+   * Return a (new) Angle object with the default "small" angle measurement (Geometry.smallAngleRadians)
```
Comment (pmconne) [inline core/geometry/src/geometry3d/Angle.ts:65]:
```suggestion
   * Return a new `Angle` object with the default "small" angle measurement specified by [[Geometry.smallAngleRadians]].
```

File: core/geometry/src/geometry3d/Angle.ts [hunk 4]
```diff
@@ -329,6 +335,14 @@ export class Angle implements BeJSONFunctions {
     const delta1 = delta - numPeriod * period;
     return Math.abs(delta1) <= radianTol;
   }
+  /**
+ * Test if this angle has magnitude less than other.
```
Comment (pmconne) [inline core/geometry/src/geometry3d/Angle.ts:339]:
```suggestion
 * Test if this angle has magnitude no greater than that of `other`.
```

File: core/geometry/src/polyface/PolyfaceQuery.ts [hunk 5]
```diff
@@ -956,35 +1028,75 @@ export class PolyfaceQuery {
...
+        result = chainContext.collectMaximalChains();
+      }
+    }
+    return result;
+  }
+  /** Find segments (within the linestring) which project to facets.
+    * * Return collected line segments
+    * * this calls  [PolyfaceQuery.sweepLineStringToFacets] with options created by
+    *   `const options = SweepLineStringToFacetsOptions.create(Vector3d.unitZ(), Angle.createSmallAngle(),false, true, true, true);`
+    * @deprecated in 4.x. Use [PolyfaceQuery.sweepLineStringToFacets] to get further options.
```
Comment (pmconne) [inline core/geometry/src/polyface/PolyfaceQuery.ts:1078]:
```suggestion
    * @deprecated in 4.x. Use [[PolyfaceQuery.sweepLineStringToFacets]] to get further options.
```

File: core/geometry/src/polyface/PolyfaceQuery.ts [hunk 6]
```diff
@@ -956,35 +1028,75 @@ export class PolyfaceQuery {
...
+    const options = SweepLineStringToFacetsOptions.create(Vector3d.unitZ(), Angle.createSmallAngle(),
+      false, true, true, true);
+    const result = PolyfaceQuery.sweepLineStringToFacets(linestringPoints, polyface, options);
+    return result as LineSegment3d[];
   }
 
   /** Find segments (within the linestring) which project to facets.
    * * Return chains.
-   * * Facets are ASSUMED to be convex and planar, and not overlap in the z direction.
+   * * this calls  [PolyfaceQuery.sweepLineStringToFacets] with options created by
```
Comment (pmconne) [inline core/geometry/src/polyface/PolyfaceQuery.ts:1089]:
```suggestion
   * * this calls  [[PolyfaceQuery.sweepLineStringToFacets]] with options created by
```

`[[ApiName]]` to link to an API in the same package.
`[ApiName]($package-alias)` to link to an API in a different package.

**Review by pmconne — APPROVED** [review]

**Review by dassaf4 — CHANGES_REQUESTED** [review]
Looks great! I pushed some doc changes and cleaned up some stuff. Please review.

**Review by dassaf4 — APPROVED** [review]

**Pull request merged**